components/common/src/main/java/com/hotels/styx/config/schema/Schema.java

```diff
@@ -137,61 +137,143 @@ public FieldValue value() {
+                    field("config", union("type"))
+            );
+
+            Optional<String> correctType = ref.isCorrectType(parents, parent, element, schemas);
+
+            if (!correctType.isPresent()) {
+                // It passed as a reference:
+                return Optional.empty();
             }
 
-            if (this instanceof MapField) {
-                return FieldType.MAP;
+            Optional<String> correctType1 = routingObjectDefinition.isCorrectType(parents, parent, element, schemas);
+            if (!correctType1.isPresent()) {
+                // It passed as a routing object spec:
+                return Optional.empty();
             }
 
-            throw new IllegalStateException("Unknown field type: " + this.getClass() + " " + this);
+            // Failed
+            return message(parents, element);
+        }
+
+        private Optional<String> message(List<String> parents, JsonNode tree) {
+            String fullName = Joiner.on(".").join(parents);
+            return Optional.of(format("Unexpected field type. Field '%s' should be %s, but it is %s", fullName, describe(), tree.getNodeType()));
+        }
+
+        @Override
+        public String describe() {
+            return "ROUTING-OBJECT-SPEC";
         }
     }
 
     /**
      * Integer schema field type.
      */
-    public static class IntegerField extends FieldValue {
+    public static class IntegerField implements FieldValue {
+        @Override
+        public Optional<String> isCorrectType(List<String> parents, JsonNode parent, JsonNode element, Function<String, FieldValue> schemas) {
+            return (element.isInt() || canParseAsInteger(element)) ? Optional.empty() : message(parents, element);
+        }
+
+        @Override
+        public String describe() {
+            return "INTEGER";
+        }
+
+        private Optional<String> message(List<String> parents, JsonNode tree) {
+            String fullName = Joiner.on(".").join(parents);
+
+            return Optional.of(format("Unexpected field type. Field '%s' should be INTEGER, but it is %s", fullName, tree.getNodeType()));
+        }
+
+        private static boolean canParseAsInteger(JsonNode value) {
+            try {
+                parseInt(value.textValue());
+                return true;
+            } catch (NumberFormatException cause) {
+                return false;
+            }
+        }
     }
 
     /**
      * String schema field type.
      */
-    public static class StringField extends FieldValue {
+    public static class StringField implements FieldValue {
+        @Override
+        public Optional<String> isCorrectType(List<String> parents, JsonNode parent, JsonNode element, Function<String, FieldValue> schemas) {
+            return element.isTextual() ? Optional.empty() : message(parents, element);
+        }
+
+        @Override
+        public String describe() {
+            return "STRING";
```

Comment:
Do we want to remove these hardcoded descriptions? Also, I don't know if we want to bring back a 'collection' of all the possible field types ... Something like the previous enum or a Kotlin sealed class.

Reply:
The "collection of all possible types" was not very accommodating for extension, and therefore had to be removed. 

This new, refactored validator framework is now open for extension. For example, `RoutingObjectSpec` could have been implemented elsewhere and simply dropped in to the validator tree. This is only possible when each validator is just a self describing class implementing a `FieldValue` interface. 

It would have been really cumbersome to achieve this with the previous "collection of all possible types" implementation.

For the same reason it is necessary for each type to describe themselves for human readable error reporting.

Reply:
TBH, there could be still better, off the shelf libraries for the validation. Ideally we'll get to deprecate this code some day.